### PR TITLE
add methods to generate random integers and strings

### PR DESCRIPTION
**Pull request by zreigz**
add methods to generate random integers and strings

### CODE REVIEW

File: algorithms/random.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+var rng = struct {
+	sync.Mutex
+	rand *rand.Rand
+}{
+	rand: rand.New(rand.NewSource(time.Now().UnixNano())),
+}
+
+// Int returns a non-negative pseudo-random int.
+func Int() int {
+	rng.Lock()
```
Comment (michaeljguarino) [inline algorithms/random.go:18]: why do we need to mutex these funcs?

File: algorithms/random.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	defer rng.Unlock()
+	return rng.rand.Int()
+}
+
+// Intn generates an integer in range [0,max).
+// By design this should panic if input is invalid, <= 0.
+func Intn(max int) int {
```
Comment (michaeljguarino) [inline algorithms/random.go:25]: we should also have that flip function, eg:

```go
func Coinflip(num, dem int) bool {
   r := Intn(dem)
   return r < num
}
```

(obviously it's not hard to do from these, but still nice to have)

File: algorithms/random.go (hunk 1 above)
Comment (zreigz) [reply algorithms/random.go:18]: To ensure that a value is not accessed concurrently. Generally, it's a thread-safe version.

File: algorithms/random.go (hunk 2 above)
Comment (zreigz) [reply algorithms/random.go:25]: done

**Review by michaeljguarino — APPROVED** [review]

File: algorithms/random.go (hunk 1 above)
Comment (michaeljguarino) [reply algorithms/random.go:18]: Oh apparently using rand.NewSource makes it un-threadsafe, weird.

**Pull request merged**